### DISCUSSION TITLE
Should the traceparent header be a string or an array of strings?

### DISCUSSION THREAD

**Discussion by pwfarwell** [open-telemetry/opentelemetry-java#3315] [Q&A]
Our application handles trace headers from incoming event messages (via a NATS pipeline).
The json for the headers looks like this, where the b3 and traceparent headers are arrays of strings:
```
  "extensions": {
    "tenantId": "KWoKXpC5NQ2bUzvB0bmaRmsK9ONRLiwB",
    "header": {
      "b3": ["89404dabe42388d7a0940f9bb76492f8-2f0a4b7aae75ed8f-1"],
      "traceparent": ["00-89404dabe42388d7a0940f9bb76492f8-2f0a4b7aae75ed8f-01"]
    }
  }
```   
To extract them, I have tried doing this 
```
  private final TextMapGetter<Map<String, List<String>>> mapGetter = new TextMapGetter<Map<String, List<String>>>() {
    
    @Override
    public List<String> get(Map<String, List<String>> carrier, String key) {
      return carrier.get(key);
    }

    // etc.
  };
```
However, the W3C and B3 implementations in java otel reject a List<String> of headers as invalid.
The only impl of a TextMapGetter that works for me is this:
```
  private final TextMapGetter<Map<String, String>> mapGetter = new TextMapGetter<Map<String, String>>() {
    
    @Override
    public String get(Map<String, String> carrier, String key) {
      return carrier.get(key);
    }

    // etc.
  };  
```
Should I be implementing the TextMapGetter differently?

**Comment by jkwatson** [comment]
I think you'll probably need to just grab the first element of the array of header values, or something like that.

**Reply by Oberon00** [reply]
I think it would be better to join by `,`. E.g. since the tracestate header is `,`-separated, I think according to the HTTP RFCs it could equivalently be represented as multiple headers. I.e., `tracestate: x=y,foo=bar` is equivalent to

```
tracestate: x=y
tracestate: foo=bar
```

**Reply by jkwatson** [reply]
That's also definitely an implementation possibility. I guess it would depend, in this case, whether the setter on the other side is doing that or not, since this isn't standard HTTP transport, as far as I can see.

**Reply by pwfarwell** [reply]
Thanks @jkwatson and @Oberon00. I tried joining two traceparent strings using a comma, but the otel W3C propagator rejects them:
```
      "traceparent": "00-ABC04dabe42388d7a0940f9bb76492f8-ABCa4b7aae75ed8f-01,00-XYZ04dabe42388d7a0940f9bb76492f8-XYZa4b7aae75ed8f-01"
```
Or are you suggesting I do something like this, with two separate headers of the same type?
```
      "traceparent": "00-ABC04dabe42388d7a0940f9bb76492f8-ABCa4b7aae75ed8f-01",
      "traceparent": "00-XYZ04dabe42388d7a0940f9bb76492f8-XYZa4b7aae75ed8f-01"
```
I can't tell from the W3C tracing spec whether it allows more than one traceparent. However, I believe it does.
For an HTTP request, does the Otel impl allow for more than one traceparent header value in a single HTTP request?

**Reply by pwfarwell** [reply]
Maybe this is my ignorance about distributed tracing.... I thought a downstream call would have only one upstream "traceparent". The upstream caller might be in a loop, or multithreaded, but each invocation of a downstream function is distinct and has only one direct parent.

What would it mean for a single "traceparent" to have an array of values?
Or does that make sense at all?

**Reply by jkwatson** [reply]
I don't think it makes any sense for there to be more than one traceparent. And, even if it did, OpenTelemetry doesn't support having a span have more than one direct parent, so we couldn't do anything with it.

**Reply by pwfarwell** [reply]
Thanks @jkwatson That makes sense to me, too. My co-workers tell me that the Go and NodeJs implementations of the otel propagators treat the ``traceparent`` and ``b3`` headers as an array of strings. :)

**Reply by jkwatson** [reply]
What do they do with multiple trace parents?

**Reply by pwfarwell** [reply]
I will try to find out...

**Reply by jkwatson** [reply]
BTW, according to spec:  https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#get

You're supposed to return the first value of the key.

**Reply by pwfarwell** [reply]
Thanks @jkwatson Thanks. So it appears that the spec allows trace headers to be provided as an array, but in practice the client should expect it to be a single string.
At any rate, the workaround of using the first value works for us.
